sw/host/opentitansession/src/main.rs

```diff
@@ -0,0 +1,263 @@
+            // `SessionStartResult` through the anonymous pipe.
+            close(wr_pipe)?;
+            let mut data = [0u8; 256];
+            let mut idx = 0;
+            loop {
+                let rc = read(rd_pipe, &mut data[idx..])?;
+                if rc == 0 {
+                    break;
+                }
+                idx += rc;
+            }
+            close(rd_pipe)?;
+            match serde_json::from_slice::<Result<SessionStartResult, String>>(&data[..idx]) {
+                Ok(Ok(result)) => {
+                    // Create a pid file corresponding to the requested TCP port.
+                    let path = run_file_fn(result.port);
+                    File::create(path)?.write_all(format!("{}\n", child).as_bytes())?;
+                    Ok(Box::new(result))
+                }
+                Ok(Err(e)) => bail!(e),
+                Err(e) => bail!("Child process failed to start: {}", e),
+            }
+        }
+        ForkResult::Child => {
+            // Child process will instantiate SessionHandler to bind to a socket, and then
+            // enter infnite loop processing connections to that socket.  The chosen port
+            // number will be reported to the parent process by means of a serialized
+            // `SessionStartResult` sent through the anonymous pipe.
+            close(rd_pipe)?;
+
+            let transport = backend::create(backend_opts)?;
+            match SessionHandler::init(&transport, listen_port) {
+                Ok(mut session) => {
+                    // Go to background, before reporting successful startup.
+                    env::set_current_dir(&Path::new("/"))?;
+                    {
+                        let devnull_file = File::open("/dev/null")?;
+                        // devnull_fd is valid only as long as devnull_file remains in scope.
+                        let devnull_fd: RawFd = devnull_file.as_raw_fd();
+                        dup2(devnull_fd, 0)?;
+                        dup2(devnull_fd, 1)?;
+                        dup2(devnull_fd, 2)?;
+                    }
+                    // Establish new Unix "session" for this process, such that SIGHUP on the
+                    // terminal that was used for running `session start` does not kill the daemon.
+                    setsid()?;
+
+                    // Report startup success to parent process.
+                    let data = serde_json::to_vec::<Result<SessionStartResult, String>>(&Ok(
+                        SessionStartResult {
+                            port: session.get_port(),
+                        },
+                    ))?;
+                    write(wr_pipe, &data)?;
+                    // Closing pipe signals that parent process can exit.
+                    close(wr_pipe)?;
+                    // Indefinitely run command processing loop in this daemon process.
+                    match session.run_loop() {
```

Comment:
This should just be `if session.run_loop().is_ok()`

Reply:
Done